### PR TITLE
Adds default http targeting context accessor

### PR DESCRIPTION
**Pull request by rossgrambo**
## Why this PR?

1. Addresses issues 
  a. https://github.com/microsoft/FeatureManagement-Dotnet/issues/415
  b. https://github.com/microsoft/FeatureManagement-Dotnet/issues/399
2. Also makes HttpContext.Items cache consistent- so all classes check the same location. 

Next Steps:
1. Remove unremarkable HttpTargetingContextAccessors from examples in favor of DefaultHttpTargetingContextAccessor.
2. Consider a different cache pattern that doesn't rely on a shared string key.

## Visible Changes

Developers will now have a new type available to them, `DefaultHttpTargetingContextAccessor`. Instead of defining their own, they can use this type when adding targeting via `WithTargeting` like:

```csharp
builder.Services.AddFeatureManagement()
    .WithTargeting<DefaultHttpTargetingContextAccessor>()
```

The default accessor will extract the targeting info from `HttpContext.User`. `Groups` will be extracted from claims of type `Role`. `UserId` is taken from the `Identity.Name` field.

### CODE REVIEW

**Comment by rossgrambo** [comment]
If someone has an idea for a shared place for the `TargetingContextLookup` string, let me know. I couldn't decide on a clean spot. Decided to leave them as separate strings for now since we're hoping to adjust the caching to not need them anyway.

**Comment by zhenlan** [comment]
Just for ideas... you either make it a public property or use InternalsVisibleToAttribute between the two packages. Like you said, none of them are clean ways.

**Comment by jimmyca15** [comment]
I know adding a scoped service will be an approach that we will want to propose to float the resolved targeting context along, but I also thing we should consider exposing a public extension method on HttpContext from the `Microsoft.FeatureManagement.AspNetCore` package that gets the resolved targeting context.

```
static class HttpContextExtensions
{
    //
    // Gets the targeting context resolved by feature management during a request.
    public static TargetingContext GetTargetingContext(this httpContext);
}
```

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:12]: namespace is wrong

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:47]: Could you add an empty line before this line?

**Comment by zhiyuanliang-ms** [comment]
@jimmyca15 @rossgrambo Are we going to include this PR into the next preview release (preview3)?

**Comment by rossgrambo** [comment]
Would be nice- but I don't think it's a blocker.

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (WeihanLi) [inline src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:17]: Maybe could be `sealed` since there's no member to override, it would be fine from `sealed` to non-sealed but breaking change from non-sealed to `sealed`

```suggestion
    public sealed class DefaultHttpTargetingContextAccessor : ITargetingContextAccessor
```

File: src/Microsoft.FeatureManagement.AspNetCore/TargetingHttpContextMiddleware.cs
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.AspNetCore/TargetingHttpContextMiddleware.cs:21]: Is there any reason why we removed `Microsoft` prefix here?

**Comment by zhiyuanliang-ms** [comment]
Is there a plan to use this built-in targeting context accessor in our FeatureFlagDemo example? (include the change in this PR or create a new PR after this PR is merged?)

File: src/Microsoft.FeatureManagement.AspNetCore/TargetingHttpContextMiddleware.cs
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.AspNetCore/TargetingHttpContextMiddleware.cs:21]: Not that I can recall... I'll add it back.

**Comment by rossgrambo** [comment]
> Is there a plan to use this built-in targeting context accessor in our FeatureFlagDemo example? (include the change in this PR or create a new PR after this PR is merged?)

Yes, I'm planning on a separate to update examples.

**Comment by jimmyca15** [comment]
There has been some developments as we have discussed the middleware and telemetry initializer across different PRs/issues. My expectation at this point is that

* The middleware moves to the same project as the initializer
  * If we stick to http context items cache, the key is project internal
* The default targeting context accessor uses a cache key that is not accessible outside
  * `private object _targetingContextKey = new object();`

As such, I don't expect any changes in the middleware/initializer as part of this PR.

**Comment by rossgrambo** [comment]
> As such, I don't expect any changes in the middleware/initializer as part of this PR.

Agreed. This PR is on hold until the caching rework is completed.

**Comment by zhiyuanliang-ms** [comment]
Open an issue: https://github.com/microsoft/FeatureManagement-Dotnet/issues/432

This PR can help resolve this issue.

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:20]:
```suggestion
        /// The key used to store and retrieve the <see cref="TargetingContext"/> from the <see cref="HttpContext"/> items.
```

**Review by zhiyuanliang-ms — APPROVED** [review]

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore.csproj
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore.csproj:40]: Is this reference still necessary?

Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore.csproj]: Good catch, no. Updated!

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs
Comment (zhenlan) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:51]: Is it intentional to let this throw if the cache doesn't exist? If so, we should throw something more meaningful to users.

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (zhenlan) [inline src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:22]: It's debatable who should own this cache key. It's odd to see the more generic initializer and middleware reference the cache key from a very specific targeting context accessor.

Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:22]: Agreed... I still am unable to come up with a satisfying place to put this. I suppose the owner could be the middleware- because the middleware's whole goal is to ensure that the `TargetingContext` is available at that key.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:51]: My undersatnding is that the `as` casting won't throw- it'll either give the requested object or return null. From https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast?redirectedfrom=MSDN#as-operator

I might be ignoring different versions or something where that's the case.

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (jimmyca15) [reply src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:22]: The targeting context accessor doesn't need to share a key.

`private static object _cacheKey = new object();` should work fine.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs
Comment (zhenlan) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:51]: I was talking about `httpContext.Items[DefaultHttpTargetingContextAccessor.TargetingContextLookup]`. Will it throw if the cache doesn't exist?

Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs]: Oh thank you- you're right. Updated

File: src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.AspNetCore/DefaultHttpTargetingContextAccessor.cs:22]: There's 3 classes that interact with HttpContext.Items for targeting context. `Initializer`, `Middleware`, and the `TargetingContextAccessor`.

`Initializer` depends upon the `Middleware`. The `Middleware` depends upon a `TargetingContextAccessor` existing. The `Middleware` and `Initializer` are tightly coupled- so they definitely need to share a key. The `TargetingContextAccessor` doesn't have to share the same key, but it reduces our HttpContext.Items footprint if it does- and is a clear shared cache location.

So for `Initializer` and `Middleware`, we are going to have a shared key- it could be either an object ref or the string key `"Microsoft.FeatureManagement.TargetingContext"`. I slightly prefer the string key since it's public (on the HttpContext.Items and as a public property). The key needs a home, and if it was just these two classes it would either be on the middleware or a shared constants class.

Now the `DefaultTargetingContextAccessor` comes in- which is our default implementation. It needs a cache key as well. It could either define it's own or used a shared constants class. 

I think the shared constants class makes things the most clear- as FM is defining where `TargetingContext` should live on `HttpContext.Items`. I ended up not liking adding the constants class because it introduces yet another public class, and I ended up going with the `DefaultTargetingContextAccessor`.

**Comment by rossgrambo** [comment]
This PR has been dissolved into two:
* https://github.com/microsoft/FeatureManagement-Dotnet/pull/466 - Add DefaultTargetingContextAccessor to `main`
* https://github.com/microsoft/FeatureManagement-Dotnet/pull/467 - Adjusts targetingId caching to use Activity

**Pull request closed without merging**
